service-base/src/main/java/org/eclipse/hono/service/command/CommandConnectionImpl.java

```diff
@@ -0,0 +1,126 @@
+                                                           final Handler<Command> commandHandler) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
+        Objects.requireNonNull(commandHandler);
+
+        LOG.debug("create a command receiver for [tenant: {}, device-id: {}]", tenantId, deviceId);
+        Future<Void> result = Future.future();
+        connect().setHandler(h -> {
+            if(h.succeeded()) {
+                getConnection().createReceiver(ResourceIdentifier.from(CommandConstants.COMMAND_ENDPOINT, tenantId, deviceId).toString()) // TODO
+                        .openHandler(oh -> {
+                            if (oh.succeeded()) {
+                                LOG.debug("command receiver successfully opened for [tenant: {}, device-id: {}]", tenantId, deviceId);
+                                ProtonReceiver protonReceiver = oh.result();
+                                CommandResponder responder = new CommandResponder(protonReceiver);
+                                protonReceiver.handler((delivery, message) -> {
+                                    LOG.debug("command message received on [address: {}]", message.getAddress());
+                                    commandHandler.handle(new Command(responder, message));
+                                });
+                                result.complete();
+                            } else {
+                                LOG.debug("command receiver failed opening for [tenant: {}, device-id: {}] : {}", tenantId, deviceId, oh.cause().getMessage());
+                                result.fail(oh.cause());
+                            }
+                        }).open();
+            }
+            else {
+                result.fail(h.cause());
+            }
+        });
+        return result;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public Future<Void> sendCommandResponse(final Command command, final Buffer data, final Map<String, Object> properties,
+                                            final Handler<ProtonDelivery> update) {
+        Objects.requireNonNull(command);
+        LOG.debug("create a command responder (sender link) for [replyAddress: {}]", command.getReplyAddress());
+        Future<Void> result = Future.future();
+        synchronized (command) {
+            ProtonSender sender = command.getResponder().getSender();
+            if (sender == null || !sender.isOpen()) {
+                connect().setHandler(h -> {
+                    getConnection().createSender(command.getReplyAddress())
+                            .openHandler(oh -> {
+                                if (oh.succeeded()) {
+                                    LOG.debug("command reply sender opened successful for [replyAddress: {}]", command.getReplyAddress());
+                                    command.getResponder().setSender(oh.result());
```

Comment:
I propose to avoid using a setter on an internal member of the command (better use a direct method on command for that to avoid changing the internals of the class...), like `setSender(...)` at the command.